libsolidity/interface/CompilerStack.cpp

```diff
@@ -103,6 +107,98 @@ CompilerStack::~CompilerStack()
 	TypeProvider::reset();
 }
 
+void CompilerStack::createAndAssignCallGraphs()
+{
+	for (Source const* source: m_sourceOrder)
+		if (source->ast)
+			for (ASTPointer<ASTNode> const& node: source->ast->nodes())
+				if (auto const* contract = dynamic_cast<ContractDefinition*>(node.get()))
+				{
+					Contract& contractState = m_contracts.at(contract->fullyQualifiedName());
+
+					contractState.contract->annotation().creationCallGraph = make_unique<CallGraph>(
+						FunctionCallGraphBuilder::buildCreationGraph(
+							*contract
+						)
+					);
+					contractState.contract->annotation().deployedCallGraph = make_unique<CallGraph>(
+						FunctionCallGraphBuilder::buildDeployedGraph(
+							*contract,
+							**contractState.contract->annotation().creationCallGraph
+						)
+					);
+
+					auto contractDeps = ranges::views::concat(
+						contract->annotation().creationCallGraph->get()->bytecodeDependency,
+						contract->annotation().deployedCallGraph->get()->bytecodeDependency
+					);
+
+					for (auto const& [dependencyContract, referencee]: contractDeps)
+					{
+						solAssert(dependencyContract != nullptr, "");
+						contractState.contract->annotation().contractDependencies.emplace(dependencyContract, referencee);
+					}
+				}
+}
+
+void CompilerStack::findAndReportCyclicContractDependencies()
+{
+	// Cycles we found, used to avoid duplicate reports for the same reference
+	map<ASTNode const*, ContractDefinition const*, ASTNode::CompareByID> foundCycles;
+
+	// Use an extra struct to remember the reference for the found cycle
+	struct RememberingCycleDetector : public util::CycleDetector<ContractDefinition>
```

Comment:
There is absolutely no need to create a subclass here. You are working with a lambda function that captures. Why don't you just use a local variable to store the reference?

Reply:
I tried many variations here and it seemed cleaner to me to have it in a subclass, but it seems I have to change this again because it's still not working as it should

Reply:
Turned out I don't need any of that. After carefully analysing the problem I found that my solution was very wrong..